### ISSUE TITLE
Parallel optimization as in original spearmint paper?

### ISSUE THREAD

**Issue by amueller** [scikit-optimize/scikit-optimize#403]
Does scikit-optimize do parallel optimization as in [the spearmint paper](https://papers.nips.cc/paper/4522-practical-bayesian-optimization-of-machine-learning-algorithms.pdf) if not, would you be interested in it?
I have a student that wants to work on auto-ml over the summer and we're brainstorming ideas. Pointers welcome.

**Comment by iaroslav-ai** [comment]
Currently "constant liar" parallelization approach is used, see eg [here](https://arxiv.org/pdf/1602.05149.pdf) (search for 'constant liar'), which is a bit different from the one in the reference. We used this approach due to its simplicity and as it can work with any of the surrogate models used to model the objective and uncertainty in skopt. We also currently support only batch parallelism. 
From my point of view, having a more advanced parallel optimization would be great, assuming it  outperforms current approach. I did not check the literature on this tbh, but for example approach in the referenced paper seems to eventually outperform constant liar. So to summarize yes on my side I think it would be great to have more better parallelization :)

**Comment by iaroslav-ai** [comment]
P.S. some other ideas that might also be interesting for brainstorming: introduce knowledge transfer, which can be used eg to speed up search for good machine learning parameters, as in [here](http://papers.nips.cc/paper/5086-multi-task-bayesian-optimization.pdf) for example. Also [scalable black box optimization](http://proceedings.mlr.press/v37/snoek15.html) with neural networks seems pretty cool, maybe something that can be used for some wild tasks like joint selection of features, neural network architecture and learning parameters.  These are some things I would also be very happy personally to see in scikit-optimize :)

**Comment by betatim** [comment]
I think it would be a nice addition. I'd also be OK with being a mentor to help them find their way around the code base.

**Comment by betatim** [comment]
(#352 is another issue on a method of parallel optimisation (qEI))

**Comment by Young-Todd** [comment]
@amueller, I am a student researcher at Oak Ridge National Lab and have been working on distributing the optimization methods in scikit-optimize on our clusters. I have a working approach that distributes smaller subregions of the hyperparameter search space for the sequential models to search over. 

https://github.com/scikit-optimize/scikit-optimize/issues/401 is related to that project where I am looking for the best way to store the many results of all of the distributed runs. 

I would be more than happy to talk with you more about what I have been working on if you are interested.

**Comment by MechCoder** [comment]
Both the strategy proposed in the paper and the one currently implemented are sequential in nature which means at least API wise, it should not be too difficult to integrate it. Here are the differences

Constant-liar.

1. Once you have the first N points, you use the SMBO loop to get a suggestion, X_{N+1}
2. To get X_{N+2}, you lie about the value at X_{N+1} to be a lie so that now you have {N+1} points and continue from 1.

We know it sucks, but as @iaroslav-ai mentions we took it as an opportunity to get the API right.

Strategy proposed in the paper:

2. To get X_{N+2}, you sample y number of values from the posterior P(Y | X_{N+1}, ((X_1, y_1), .... (X_N, y_N)). Once you have these samples, you can compute the expected acquisition function using MC approximations and then return X_{N+2} that maximizes this.

Disadvantages:

1. For each of the y values sampled from the posterior, you would have to fit a new GP and compute the acquisition function value at these points which is expensive.
2. How to optimize? I would assume they would use a fine-grid to optimize the expected acquisition function and not use any numerical-gradient based techniques since this function itself is not too cheap.

**Comment by MechCoder** [comment]
@Young-Todd Your approach is definitely interesting, feel free to open a new issue or a pull request in the form of a notebook if you want to demonstrate how to use scikit-optimize to distribute the search-space among different Optimizer or even more directly into the Optimizer API if you feel that is possible.

This is just so that the attention does not get distributed (no pun intended) :-)

**Comment by betatim** [comment]
@MechCoder (you seem to know these things) How does q-EI (see for example https://arxiv.org/pdf/1602.05149.pdf but don't think this is the original qEI paper) fit into these different methods?

**Comment by MechCoder** [comment]
q-EI is not sequential in nature. It is an extension of EI to handle multiple points.

`EI(x) = - E(min(f(x) - f*, 0))`, it gives a reward proportional to how much `f(x)` is smaller than the previous minimum `f*`

`qEI(x1, x2, ..... xN) = -E(min(t(x1, x2, .. xN) - f*, 0))` where `t(x1, x2, .. xN)` equals `min(f(x1), f(x2) ... f(xN))`, it gives a reward proportional to how much the minimum of the N points is smaller than the previous minimum `f*`

You can obtain closed-form solution for EI and thus have access to gradients while in q-EI you can only make MC approximations from the joint posterior distribution of P(f(x1), f(x2) .. f(xN))

**Comment by amueller** [comment]
So should we start with qEl or the Spearmint method? You say it's sequential in nature. What do you mean by that? Is there any comparison of this with qEl?

**Comment by betatim** [comment]
I have no opinion on qEI vs spearmint method, just wanted mentioned it.

**Comment by MechCoder** [comment]
By sequential, I mean that while suggesting n points, in a sequential method for suggesting the second point, you would need the first point, (for example, the algorithm currently implemented)

For q-EI, you generate the `n` points at one go. This is because this is a special acquisition function that takes n number of points as arguments and returns 1 value.

**Comment by amueller** [comment]
Well in the spearmint method you generate the points one-by-one from what I remember, but you don't need the actual values, because you integrate them out, right?

**Comment by iaroslav-ai** [comment]
I think what @MechCoder was trying to say is that because acquisition is optimized over a single point in Spearmint method, it is easier to implement with current API.

I also think so, thus my suggestion would be to start with Spearmint method, for which no optimization over multiple points necessary. I see it as an extension of Constant Liar strategy, where instead of a single lie, multiple lies are generated (ideally all possible, but this is not feasible of course) and then outputs of different surrogates are averaged. So it could work like this. 

1. Generate m combinations of objective assignment to n unlabeled points, using eg mean and std of surrogate fitted on the "labeled" points
2. Use only these m instances of (y_1, y_2, ...) for approximation of integral during the optimization of acquisition function (it is fixed during generation of new point). Calculate probabilities of different (y_1, y_2, ...) using the mean and std obtained from "supervised" surrogate.

Then a Constant Liar of some flavor is just a special instance of above when m = 1.

**Comment by kejiashi** [comment]
@betatim Hi Tim, I'm @amueller 's student working on this project. Currently moving on to study skopt's code. I would update my progress in this post.

**Comment by betatim** [comment]
Welcome 👋

I find a good way to get into the code base is to solve some of the issues marked as "easy" before starting on something big. Also open a PR as soon as possible instead of waiting till you think it is perfect. Having small chunks of code to look at and see how it grows makes it much easier to review than one 200 line diff :)

**Comment by kejiashi** [comment]
Thanks, Tim. Gotcha! I'll try as you said.

**Comment by kejiashi** [comment]
I'm having some doubts on where to add the sampling parameters. Shall I update the `Optimizer` class  `acq_optimizer_kwargs` with two more sampling-related parameters? (As demonstrated in this commit.) What's the best position to add this?

Also should I change the name of `sampling` (as the original sampling is point wise max likelihood sampling), or I add `mcmc_slice` as a choice under `sampling`?
